include/mls/treekem.h

```diff
@@ -179,6 +179,14 @@ struct TreeKEMPublicKey
     const FilteredDirectPath& fdp,
     const std::vector<UpdatePathNode>& path_nodes) const;
 
+  using TreeHashCache = std::map<std::tuple<NodeIndex, size_t>, bytes>;
```

Comment:
Just to confirm I understand, this works because the parent hash algorithm traverses the tree in "one direction", in the sense of the `E1 >= E2 >= ... >= Ek` chain above?  So the cache captures the cases where `Ei = ... = Ej` for some span.  Want to capture any dependency between the caching strategy and the traversal to avoid breaking one by changing the other.

Otherwise, this doesn't seem too intimdating, will give it a try.